contracts/Marketplace.sol

```diff
@@ -1,74 +1,272 @@
-        sid = _sid;
-        transferOwnership(creator);
-        addPauser(creator);
-        renouncePauser();
-    }
+contract Marketplace is Ownable, Pausable {
+  // ------------------------------------------------------
+  // Structures
+  // ------------------------------------------------------
 
-    function createVersion(bytes32 versionHash, bytes32 location, uint256 price) public onlyOwner whenNotPaused {
-        require(!versions[versionHash]._exists, "This version already exists");
-        versions[versionHash] = Version({
-            id: versionHash,
-            location: location,
-            price: price,
-            _exists: true
-        });
-        latest = versionHash;
-        emit VersionCreated(sid, versionHash, location, price);
-        return;
-    }
+  struct Service {
+    address owner;
+    string sid;
+    Version[] versions;
+    Offer[] offers;
+    Payment[] payments;
+  }
+
+  struct Version {
+    string hash;
+    string url;
+  }
+
+  struct Offer {
+    uint price;
+    // address payment; // TODO: to implement later
+    address payable seller;
+    bool active;
+  }
+
+  struct Payment {
+    uint offerIndex;
+    address purchaser;
+  }
+
+  // ------------------------------------------------------
+  // TODO: Events
+  // ------------------------------------------------------
+
+  // ------------------------------------------------------
+  // State variables
+  // ------------------------------------------------------
+
+  Service[] public services; //TODO: shouldn't it be private?
+
+  // ------------------------------------------------------
+  // Constructor
+  // ------------------------------------------------------
+
+  constructor() public {}
 
-    function requestAccess(bytes32 version) public whenNotPaused {
-        require(!_versionAccess[version][msg.sender], "You already have access to this version of the service");
-        _versionAccess[version][msg.sender] = true;
-        return;
+  // ------------------------------------------------------
+  // Check functions
+  // ------------------------------------------------------
+
+  function checkServiceOwner(Service memory service) private view {
+    require(service.owner == msg.sender, "Service owner is not the same as the sender");
+  }
+
+  // ------------------------------------------------------
+  // Utils functions
+  // ------------------------------------------------------
+
+  function compare(string memory a, string memory b) internal pure returns (bool) {
+    return keccak256(bytes(a)) == keccak256(bytes(b));
+  }
+
+  // ------------------------------------------------------
+  // View functions
+  // ------------------------------------------------------
+
+  // Manage Service
+
+  function getServiceIndexRaw(string memory sid) public view returns (int serviceIndex) {
+    for (uint i = 0; i < services.length; i++) {
+      if (compare(services[i].sid, sid)) {
+        return int(i);
```

Comment:
I don't like casting index of array - this should be uint always

Reply:
Improved